### PR TITLE
Reorganization

### PR DESCRIPTION
**Pull request by alexkart**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Tests pass?   | ✔️
| Fixed issues  | #29 

Old code moved to `src-old` and `tests-old`, new code is in `src` and `tests`.
Only `Memcached` and `ArrayCache` implementations are rewritten to the new design for now.
There are also some doc blocks and type hints may be missing yet.

### CODE REVIEW

**Comment by rob006** [comment]
> Old code moved to `src-old` and `tests-old`

Why? This makes review much more difficult...

**Comment by alexkart** [comment]
> > Old code moved to `src-old` and `tests-old`
> 
> Why? This makes review much more difficult...

I've started from scratch, so, there is no sense in looking at the diff, the best way is to checkout PR and review the new code.

**Comment by samdark** [comment]
It is better to replace current files to get the diff...

File: src/ArrayCache.php [hunk 1]
```diff
@@ -1,49 +1,175 @@
...
- */
-final class ArrayCache extends SimpleCache
+final class ArrayCache extends BaseCache
 {
-    private const TTL_INFINITY = 0;
+    private $cache = [];
 
     /**
-     * @var array cached values.
+     * Fetches a value from the cache.
```
Comment (samdark) [inline src/ArrayCache.php]: There's no need to duplicate interface phpdoc.

File: src/ArrayCache.php [hunk 2]
```diff
@@ -1,49 +1,175 @@
...
-    private $cache = [];
-
-    protected function hasValue(string $key): bool
+    public function get($key, $default = null)
     {
-        return isset($this->cache[$key]) && ($this->cache[$key][1] === 0 || $this->cache[$key][1] > time());
-    }
+        if (isset($this->cache[$key]) && !$this->isExpired($key)) {
+            $value = $this->cache[$key][0];
+            if (is_object($value)) {
```
Comment (samdark) [inline src/ArrayCache.php]: Good catch. `clone` makes perfect sense there 👍

File: src/BaseCache.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+abstract class BaseCache implements CacheInterface
+{
+    public const EXPIRATION_INFINITY = 0;
+    /**
+     * @var int|null default TTL for a cache entry. null meaning infinity, negative or zero results in cache key deletion.
+     * This value is used by {@see set()} and {@see setMultiple()}, if the duration is not explicitly given.
+     */
+    private $defaultTtl;
+
+    public function __construct()
```
Comment (samdark) [inline src/BaseCache.php]: Should be removed since it does nothing.

File: src/BaseCache.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+     * @throws \Exception
+     */
+    protected function ttlToExpiration($ttl): int
+    {
+        $ttl = $this->normalizeTtl($ttl);
+
+        if ($ttl === null) {
+            $expiration = static::EXPIRATION_INFINITY;
+        } elseif ($ttl <= 0) {
+            $expiration = -1;
```
Comment (samdark) [inline src/BaseCache.php]: `-1` could be extracted into `EXPIRATION_EXPIRED` or `EXPIRATION_DELETE`.

File: src/Cache.php [hunk 5]
```diff
@@ -55,38 +73,39 @@ public function __construct(\Psr\SimpleCache\CacheInterface $handler = null)
...
+        if (!$this->keyNormalization) {
+            $normalizedKey = $key;
+        } elseif (\is_string($key)) {
+            $normalizedKey = ctype_alnum($key) && mb_strlen($key, '8bit') <= 32 ? $key : md5($key);
+        } else {
+            $normalizedKey = $this->keyPrefix . md5(json_encode($key));
         }
-        return md5(json_encode($key));
+
+        return $this->keyPrefix . $normalizedKey;
```
Comment (samdark) [inline src/Cache.php]: Normalization and prefixing are two separate actions. `normalizeKey()` name suggests that normalization is done inside so either it's not a good name (`buildKey()` is fine in this case) or there should be two methods.

File: src/Cache.php [hunk 6]
```diff
@@ -55,38 +73,39 @@ public function __construct(\Psr\SimpleCache\CacheInterface $handler = null)
...
      */
-    private function buildKey($key): string
+    private function normalizeKey($key): string
     {
-        if (\is_string($key)) {
-            return ctype_alnum($key) && mb_strlen($key, '8bit') <= 32 ? $key : md5($key);
+        if (!$this->keyNormalization) {
+            $normalizedKey = $key;
+        } elseif (\is_string($key)) {
+            $normalizedKey = ctype_alnum($key) && mb_strlen($key, '8bit') <= 32 ? $key : md5($key);
```
Comment (samdark) [inline src/Cache.php]: Not sure these checks are faster then just `json_encode` everything.

File: src/Cache.php [hunk 7]
```diff
@@ -104,15 +123,16 @@ public function has($key): bool
      */
     public function getMultiple($keys, $default = null): iterable
     {
+        // TODO refactor
```
Comment (samdark) [inline src/Cache.php]: ⚠️

File: src/Cache.php [hunk 8]
```diff
@@ -104,15 +123,16 @@ public function has($key): bool
...
-            $keyMap[$key] = $this->buildKey($key);
+            $keyMap[$key] = $this->normalizeKey($key);
         }
-        $values = $this->handler->getMultiple(array_values($keyMap));
+        $values = $this->handler->getMultiple(array_values($keyMap), $default);
         $results = [];
         foreach ($keyMap as $key => $newKey) {
             $results[$key] = $default;
-            if (isset($values[$newKey])) {
+            if (array_key_exists($newKey, (array)$values)) {
```
Comment (samdark) [inline src/Cache.php]: `getMultiple` can not return non-array.

File: src/Cache.php [hunk 9]
```diff
@@ -313,4 +340,68 @@ public function getOrSet($key, callable $callable, $ttl = null, Dependency $depe
 
         return $value;
     }
+
+    public function __call($name, $arguments)
```
Comment (samdark) [inline src/Cache.php]: What's the purpose for it?

File: src/ApcuCache.php [hunk 10]
```diff
@@ -1,56 +0,0 @@
...
-
-/**
- * ApcuCache provides APCu caching in terms of an application component.
- *
- * To use this application component, the [APCu PHP extension](http://www.php.net/apcu) must be loaded.
- * In order to enable APCu for CLI you should add "apc.enable_cli = 1" to your php.ini.
- *
- * See {@see \Psr\SimpleCache\CacheInterface} for common cache operations that ApcCache supports.
- */
-final class ApcuCache extends SimpleCache
```
Comment (samdark) [inline src/ApcuCache.php]: Why removing APCu?

File: src/Cache.php [hunk 11]
```diff
@@ -313,4 +340,68 @@ public function getOrSet($key, callable $callable, $ttl = null, Dependency $depe
...
+
+    /**
+     * @param string $keyPrefix a string prefixed to every cache key so that it is unique globally in the whole cache storage.
+     * It is recommended that you set a unique cache key prefix for each application if the same cache
+     * storage is being used by different applications.
+     */
+    public function setKeyPrefix(string $keyPrefix): void
+    {
+        if ($keyPrefix != '' && !ctype_alnum($keyPrefix)) {
+            throw new Exception\InvalidArgumentException('Cache key prefix should be alphanumeric');
```
Comment (samdark) [inline src/Cache.php]: That rule disables `-`, `_` and `.` that are common for cache keys.

File: src/Cache.php [hunk 12]
```diff
@@ -37,12 +40,27 @@ final class Cache implements CacheInterface
...
+
+    private $keyNormalization = true;
+
     /**
      * @param \Psr\SimpleCache\CacheInterface cache handler.
      */
     public function __construct(\Psr\SimpleCache\CacheInterface $handler = null)
     {
         $this->handler = $handler;
+        $this->initSerializer();
```
Comment (samdark) [inline src/Cache.php]: It's one liner that is clear about what it does. There is no need for separate method.

File: src/FileCache.php [hunk 13]
```diff
@@ -1,266 +0,0 @@
...
-/**
- * FileCache implements a cache handler using files.
- *
- * For each data value being cached, FileCache will store it in a separate file.
- * The cache files are placed under {@see FileCache::$cachePath}. FileCache will perform garbage collection
- * automatically to remove expired cache files.
- *
- * Please refer to {@see \Psr\SimpleCache\CacheInterface} for common cache operations that are supported by FileCache.
- */
-final class FileCache extends SimpleCache
```
Comment (samdark) [inline src/FileCache.php]: Why removing file cache?

File: src/Memcached.php [hunk 14]
```diff
@@ -2,252 +2,180 @@
...
 
-            $this->cache = $this->persistentId !== null ? new \Memcached($this->persistentId) : new \Memcached();
-            if ($this->username !== null || $this->password !== null) {
-                $this->cache->setOption(\Memcached::OPT_BINARY_PROTOCOL, true);
-                $this->cache->setSaslAuthData($this->username, $this->password);
-            }
-            if (!empty($this->options)) {
-                $this->cache->setOptions($this->options);
-            }
+        // TODO store error message
```
Comment (samdark) [inline src/Memcached.php]: ⚠️

File: src/NullCache.php [hunk 15]
```diff
@@ -33,7 +35,7 @@ public function get($key, $default = null)
 
     public function getMultiple($keys, $default = null): iterable
     {
-        return array_fill_keys($keys, $default);
+        return array_fill_keys((array)$keys, $default);
```
Comment (samdark) [inline src/NullCache.php]: Keys should always be array. Not sure we need such type-cast.

File: src/WinCache.php [hunk 16]
```diff
@@ -1,63 +0,0 @@
...
-
-/**
- * WinCache provides Windows Cache caching in terms of an application component.
- *
- * To use this application component, the [WinCache PHP extension](https://sourceforge.net/projects/wincache/)
- * must be loaded. Also note that "wincache.ucenabled" should be set to "1" in your php.ini file.
- *
- * See {@see \Psr\SimpleCache\CacheInterface} for common cache operations that are supported by WinCache.
- */
-final class WinCache extends SimpleCache
```
Comment (samdark) [inline src/WinCache.php]: Why removing it?

File: tests/BaseTest.php [hunk 17]
```diff
@@ -0,0 +1,308 @@
...
+
+    public function testGetNonExistent(): void
+    {
+        $cache = $this->createCacheInstance();
+        $cache->clear();
+
+        $this->assertNull($cache->get('non_existent_key'));
+    }
+
+    // TODO
```
Comment (samdark) [inline tests/BaseTest.php]: ⚠️

**Comment by samdark** [comment]
Architecture changes as I see it.

1. It became cleaner implementation-wise with the fact that the same interface is used.
2. Not sure why default TTL was left at the PSR implementation level but key prefixing, key normalization and serialization were moved to `Cache` decorator. There are PSR-16 implementations that do not contain default TTL implementation but contain serialization implementation such as https://github.com/kodedphp/cache-simple.
3. Because of moving some of the configuration to `Cache` decorator, it's now a bit harder to configure cache for adapters that are not our own. You have to read source code of third party implementation and decide on key prefixing, key normalization and serialization. Previously it was not required as these features were present in Yii implementations only.

**Comment by samdark** [comment]
@rob006 what do you think?

File: .travis.yml [hunk 18]
```diff
@@ -35,8 +35,8 @@ install:
   # setup PHP extension
   - |
     if [[ $TRAVIS_PHP_VERSION != nightly ]]; then
-      tests/travis/apcu-setup.sh
-      tests/travis/memcache-setup.sh
+      source tests/travis/apcu-setup.sh
+      source tests/travis/memcache-setup.sh
```
[the remark this answers is not in the document]
Comment (alexkart) [reply .travis.yml]: For some reason it doesn't work with my travis account, I get permission denied error.

File: .travis.yml [hunk 19]
```diff
@@ -45,7 +45,7 @@ before_script:
   - composer --version
 
 script:
-  - phpdbg -qrr vendor/bin/phpunit --verbose --coverage-clover=coverage.clover --exclude-group wincache
+  - vendor/bin/phpunit --verbose --coverage-clover=coverage.clover --exclude-group wincache
```
[the remark this answers is not in the document]
Comment (alexkart) [reply .travis.yml]: It doesn't work in my travis account

File: src/ApcuCache.php (hunk 10 above)
Comment (alexkart) [reply src/ApcuCache.php]: As I mentioned it's not rewritten yet.

File: src/ArrayCache.php (hunk 1 above)
Comment (alexkart) [reply src/ArrayCache.php]: Agree

File: src/FileCache.php (hunk 13 above)
Comment (alexkart) [reply src/FileCache.php]: As I mentioned it's not rewritten yet.

File: src/WinCache.php (hunk 16 above)
Comment (alexkart) [reply src/WinCache.php]: As I mentioned it's not rewritten yet.

File: src/NullCache.php (hunk 15 above)
Comment (alexkart) [reply src/NullCache.php]: Interface says it's iterable

**Comment by alexkart** [comment]
> Not sure why default TTL was left at the PSR implementation level but key prefixing, key normalization and serialization were moved to Cache decorator.

I also wasn't sure where it should be. I can try to move it to the `Cache`.

File: src/Memcached.php (hunk 14 above)
Comment (alexkart) [reply src/Memcached.php]: Do you think it can be useful or there is no need for that?

Comment (samdark) [reply src/Memcached.php]: I think there's no need for it.

File: src/NullCache.php (hunk 15 above)
Comment (samdark) [reply src/NullCache.php]: Ah, correct. Are you sure type-casting works well? There's https://www.php.net/manual/en/function.iterator-to-array.php

**Comment by rob006** [comment]
> @rob006 what do you think?

I don't really see how this simplifies anything, while it definitely makes this library more tricky and harder to use. Serialization and keys normalization should be backend specific - different backends may have different rules for keys and may need (or don't) serialization. 

How `FileCache` is supposed to work with this design? Since serialization is moved to `Cache` decorator, does it mean that if I try to use `$fileCache->set('my key', [])` it will throw error because it will not be able to save array to file?

**Comment by alexkart** [comment]
> How FileCache is supposed to work with this design? Since serialization is moved to Cache decorator, does it mean that if I try to use $fileCache->set('my key', []) it will throw error because it will not be able to save array to file?

It's an interesting question. I think those are different layers of responsibility. For example, besides serialization that provided by `Cache` php `\Memcached()` library has its own internal storage logic and serialization/deserialization to provide the ability to store all php types. So, `FileCache` implementation should also use the same serializers fro the library but only for non-supported types.
1. You will have the ability to use `FileCache` without decorator (strings will be saved as is without serializations, but other types will be serialized/deserialized)
2. You will have the ability to use `Cache` decorator with serializer with `FileCache` (in this case all data types including strings will be serialized by `Cache` and serializer of the `FileCache` will not be applied as all pieces of data that are passed to the `FileCache` are already strings and strings are saved as is by the `FileCache`).

It's quite flexible. It looks like some sort of logic duplication but I think it's not because as I said it's different layers of responsibility.
`FileCache` responsibility is to know how to store and read data.
`Cache` responsibility is how to transform or decorate data that will be passed to the cache service implementation.

**Comment by rob006** [comment]
> `FileCache` responsibility is to know how to store and read data.
> `Cache` responsibility is how to transform or decorate data that will be passed to the cache service implementation.

From: https://www.php-fig.org/psr/psr-16/#14-data:
> Implementing libraries MUST support all serializable PHP data types

So `FileCache` which does not handle serialization of arrays is incorrect implementation of PSR interface.

**Comment by alexkart** [comment]
Yes, `FileCache` must know how to save all types, I said exactly that.

**Comment by rob006** [comment]
How do you want to achieve that without serialization at `FileCache` level? A thought this was an `Cache` decorator responsibility.

**Comment by alexkart** [comment]
It will use serialization for types that can't be saved as is. I mentioned it my comment above

**Comment by samdark** [comment]
If PSR-16 must implement serialization, why do we need serialization at Cache decorator level?

**Comment by rob006** [comment]
> It will use serialization for types that can't be saved as is. I mentioned it my comment above

How did you distinguish between cached `null` and `N;` string?

**Comment by alexkart** [comment]
> How do you distinguish between cached null and N; string?

You are right, partial serialization is not an option.

> If PSR-16 must implement serialization, why do we need serialization at Cache decorator level?

Agree, serialization is an internal concern of the PSR implementation class. So, we don’t need it on the Cache level. But it shouldn’t be in a common class, serialization staff needs to be implemented in each PSR implementation class, which needs it.
One more thought about common class (like SimpleCache or BaseCache in my version) - I see that you also plan to extract each PSR implementation into separate package, as far as I understand there can’t be any parent classes in this case the because you would need to duplicate it in each repository.
So, each backend implementation needs to handle ttl normalization, in addition to that some of them (FileCache) need to handle serialization. Is there a case when it’s required to use serialization for Memcached and ArrayCache implementations. As they work perfectly with all types without serialization?
In this case each backend specific package can be used separately as pure PSR implementation or the main package with all backends and the decorator can be used as an enhanced cache library.

**Comment by samdark** [comment]
Logically correct.

We'll have to duplicate some code in case we are going to merge `SimpleCache` into each implementation but for some backends that should reduce complexity and amount of code. Considering further separation it makes sense to do.

**Comment by samdark** [comment]
Consider https://github.com/yiisoft/yii-web/issues/70

File: src/NullCache.php (hunk 15 above)
Comment (alexkart) [reply src/NullCache.php]: Yes, casting for an array works correctly for all iterables except `IteratorAggregate` and generators.
I think we need to use this:
```php
$array = $iterable instanceof \Traversable ? iterator_to_array($iterable) : $iterable;
```

Comment (samdark) [reply src/NullCache.php]: Or just `iterator_to_array` all the time.

Comment (alexkart) [reply src/NullCache.php]: It will fail with array `iterator_to_array([])` it only accepts Traversable

**Comment by alexkart** [comment]
> Consider [yiisoft/yii-web#70](https://github.com/yiisoft/yii-web/issues/70) when deciding to remove serializer from backend. There could be more usages for it than allowing complicated values.

Do you mean each backend should include serializer?
I've looked through the discussion in https://github.com/yiisoft/yii-web/issues/70 and I also agree with the opinion that encryption is an application responsibility. So, I don't see the point to include it to each PSR implementation. 
We could leave serializer support at Cache level, so that it can be added  to any cache implementation (**it's an additional feature**) and some cache backends (like `FileCache`) will also use an internal serializer as it's not possible to implement it without serialization (**in this case serializer is an implementation requirement** for some cache backends).

**Comment by samdark** [comment]
Thinking about it more, if both Cache decorator and individual implementations are PSR-16, there could be PSR-16 decorator that encrypts and decrypts data so yes. No need for serialization support for each backend.

File: src/Cache.php (hunk 6 above)
Comment (alexkart) [reply src/Cache.php]: This logic was there before.

File: src/Cache.php (hunk 8 above)
Comment (alexkart) [reply src/Cache.php]: It returns `iterable` I'll use `iterableToArray` here like in other places.

File: src/Cache.php (hunk 9 above)
Comment (alexkart) [reply src/Cache.php]: Some cache backends can have specific public methods (for example, `Memcached` implementation has `addServer()` method) we can't call that method if we use decorator with `Memcached` class.

File: src/Cache.php (hunk 11 above)
Comment (alexkart) [reply src/Cache.php]: This code was there before, I didn't add it, should I remove it?

File: src/Cache.php (hunk 9 above)
Comment (samdark) [reply src/Cache.php]: These methods are not meant to work runtime. There are more about configuration intended to be called before decorator is applied:

```php
$memcacheBackend = new Memcached();
$memcacheBackend->addServer(...);
$cache = new Cache($memcacheBackend);
```

I see no need to proxy these. Let's remove `__call()`.

File: src/Cache.php (hunk 11 above)
Comment (samdark) [reply src/Cache.php]: No.

File: src/Cache.php (hunk 9 above)
Comment (alexkart) [reply src/Cache.php]: Yes, it's only for configuration. I'll remove it from `Cache`. 
But I'm thinking of adding `__call()` to the `Memcached` class. PHP `\Memcached` library has several methods related to configuration:
```
addServer
addServers
getOption
getServerByKey
getServerList
getStats
getVersion
isPersistent
isPristine
resetServerList
setOption
setOptions
setSaslAuthData
```
So, with `__call` method we don't need to add proxy method for each of those. What do you think?

Comment (alexkart) [reply src/Cache.php]: Another option is to just expose `\Memcached` instance via getter `getCache`.

Comment (samdark) [reply src/Cache.php]: I don't like magic.

Comment (alexkart) [reply src/Cache.php]: What about getter?

Comment (samdark) [reply src/Cache.php]: That's OK.

Comment (rob006) [reply src/Cache.php]: Please do not use `__call()` when you can avoid it - it just adds another layer of magic and makes code harder to understand and debug.

File: src/ArrayCache.php [hunk 20]
```diff
@@ -1,41 +1,51 @@
...
 /**
  * ArrayCache provides caching for the current request only by storing the values in an array.
  *
  * See {@see \Psr\SimpleCache\CacheInterface} for common cache operations that ArrayCache supports.
  */
-final class ArrayCache extends SimpleCache
+final class ArrayCache implements CacheInterface
 {
-    private const TTL_INFINITY = 0;
+    public const EXPIRATION_INFINITY = 0;
```
Comment (samdark) [inline src/ArrayCache.php]: Is it used outside of the class?

File: src/Memcached.php [hunk 21]
```diff
@@ -2,252 +2,235 @@
...
  * All data in memcached can be accessed by any process running in the system.
- *
- * You can configure more properties of each server, such as `persistent`, `weight`, `timeout`.
- * Please see {@see MemcachedServer} for available options.
  */
-final class Memcached extends SimpleCache
+final class Memcached implements CacheInterface
 {
-    private const TTL_INFINITY = 0;
+    public const EXPIRATION_INFINITY = 0;
```
Comment (samdark) [inline src/Memcached.php]: Are these used outside of the class?

File: src/ArrayCache.php (hunk 20 above)
Comment (alexkart) [reply src/ArrayCache.php]: No, should be private.

File: .travis.yml (hunk 18 above)
[the remark this answers is not in the document]
Comment (alexkart) [reply .travis.yml]: It's strange, it doesn't work now in yiisoft travis either https://travis-ci.org/yiisoft/cache/jobs/570545663

**Comment by alexkart** [comment]
I've added all remaining cache backend implementations. They are ready to be extracted to separate repositories.

File: src/ApcuCache.php [hunk 22]
```diff
@@ -9,48 +15,83 @@
...
+     * Normalizes cache TTL handling `null` value and {@see DateInterval} objects.
+     * @param int|DateInterval|null $ttl raw TTL.
+     * @return int|null TTL value as UNIX timestamp or null meaning infinity
+     */
+    private function normalizeTtl($ttl): ?int
+    {
+        $normalizedTtl = $ttl;
+        if ($ttl instanceof DateInterval) {
+            try {
+                $normalizedTtl = (new DateTime('@0'))->add($ttl)->getTimestamp();
```
Comment (samdark) [inline src/ApcuCache.php]: When would it throw an exception?

File: src/ArrayCache.php [hunk 23]
```diff
@@ -46,4 +58,83 @@ public function clear(): bool
...
+    {
+        return isset($this->cache[$key]) && !$this->isExpired($key);
+    }
+
+    /**
+     * Checks whether item is expired or not
+     * @param $key
+     * @return bool
+     */
+    private function isExpired($key): bool
```
Comment (samdark) [inline src/ArrayCache.php]:
```suggestion
    private function isExpired(string $key): bool
```

Should we type-hint it?

File: src/ArrayCache.php [hunk 24]
```diff
@@ -46,4 +58,83 @@ public function clear(): bool
...
+    {
+        return $this->cache[$key][1] !== 0 && $this->cache[$key][1] <= time();
+    }
+
+    /**
+     * Converts TTL to expiration
+     * @param $ttl
```
Comment (samdark) [inline src/ArrayCache.php]:
```suggestion
     * @param int|DateInterval|null $ttl
```

File: src/FileCache.php [hunk 25]
```diff
@@ -158,23 +113,111 @@ protected function setValue(string $key, $value, ?int $ttl): bool
...
+    }
+
+    public function has($key)
+    {
+        return $this->existsAndNotExpired($key);
+    }
+
+    /**
+     * Converts TTL to expiration
+     * @param $ttl
```
Comment (samdark) [inline src/FileCache.php]:
```suggestion
     * @param int|DateInterval|null $ttl
```

File: src/Memcached.php [hunk 26]
```diff
@@ -2,252 +2,241 @@
...
     {
-        $this->username = $username;
+        $this->cache = new \Memcached($this->persistentId);
     }
 
     /**
-     * @param string $password memcached sasl password.
-     * @see http://php.net/manual/en/memcached.setsaslauthdata.php
+     * Converts TTL to expiration
+     * @param $ttl
```
Comment (samdark) [inline src/Memcached.php]:
```suggestion
     * @param int|DateInterval|null $ttl
```

File: src/Memcached.php [hunk 27]
```diff
@@ -2,252 +2,241 @@
...
+     */
+    private function normalizeTtl($ttl): ?int
     {
-        $values = $this->cache->getMulti($keys);
-
-        if ($this->cache->getResultCode() === \Memcached::RES_SUCCESS) {
-            return $values;
+        if ($ttl instanceof DateInterval) {
+            try {
+                return (new DateTime('@0'))->add($ttl)->getTimestamp();
```
Comment (samdark) [inline src/Memcached.php]: When would it throw?

File: src/Memcached.php [hunk 28]
```diff
@@ -2,252 +2,241 @@
...
+    /**
+     * @param array $servers
+     */
+    private function initServers(array $servers): void
     {
-        if ($ttl === null) {
-            $ttl = self::TTL_INFINITY;
-        } else {
-            $ttl += time();
+        if (empty($servers)) {
```
Comment (samdark) [inline src/Memcached.php]:
```suggestion
        if ($servers === []) {
```

File: src/NullCache.php [hunk 29]
```diff
@@ -65,4 +67,21 @@ public function has($key): bool
     {
         return false;
     }
+
+    public function enableKeyNormalization(): void
+    {
```
Comment (samdark) [inline src/NullCache.php]: I'd add the following comment to make intent clear:

```php
// do nothing
```

File: src/NullCache.php [hunk 30]
```diff
@@ -65,4 +67,21 @@ public function has($key): bool
...
+    }
+
+    public function disableKeyNormalization(): void
+    {
```
Comment (samdark) [inline src/NullCache.php]: I'd add the following comment to make intent clear:

```php
// do nothing
```

File: src/NullCache.php [hunk 31]
```diff
@@ -65,4 +67,21 @@ public function has($key): bool
...
+    }
+
+    public function setKeyPrefix(string $keyPrefix): void
+    {
```
Comment (samdark) [inline src/NullCache.php]: I'd add the following comment to make intent clear:

```php
// do nothing
```

**Comment by samdark** [comment]
Overall looks great! We need to decide on dependencies by deciding on should individual drivers depend on yiisoft/cache or not. If not, using exceptions from yiisoft/cache is not a good idea.

**Comment by alexkart** [comment]
I think we need to implement `InvalidArgumentException` and `CacheException` in each driver as PSR-16 requires them.

File: src/Memcached.php (hunk 27 above)
Comment (alexkart) [reply src/Memcached.php]: I think never, as we pass constant to `DateTime`, but it's declared in the `DateTime` doc block that it throws an exception and if don't catch it in the `normalizeTtl` we will have to specify `@throws Exception` tag also.

Comment (rob006) [reply src/Memcached.php]: > Do you know a better way to deal with this?

What about doing nothing? This is only false-positive from inspection.

**Comment by rob006** [comment]
What is wrong with having `yiisoft/cache` as a dependency? Having some base class and exceptions will reduce code duplication.

**Comment by alexkart** [comment]
> What is wrong with having `yiisoft/cache` as a dependency? Having some base class and exceptions will reduce code duplication.

Technically nothing wrong but in this case `yiisoft/cache` will contain `Cache` class the main thing of the library and additionally another class that is used only by another package and which is have nothing to do with the main `Cache` class. It looks a bit strange. The package will have the second unclear responsibility.

**Comment by rob006** [comment]
> The package will have the second unclear responsibility.

What second responsibility? I thought this library will contain dependency-free backends (like file or array cache). Separate packages will be only for implementations which requires additional dependencies (like redis).

File: src/Memcached.php (hunk 27 above)
Comment (samdark) [reply src/Memcached.php]: I'd just remove it and let it complain or use special IDE annotations such as `/** @noinspection PhpDocMissingThrowsInspection */` to make it silent.

**Comment by alexkart** [comment]
My understanding that it will be the following:

- `cache` package with `Cache` decorator, dependencies
- `memcached` package with Memcached implementation
- `file-cache` package with file cache implementation
etc.
All packages can be used separately or in any combinations.
For example,

- you can just use `memcached` package as Memcached PSR implementation
- you can just use `memcached` package with `cache` package for additional functionality
- you can use `cache` package with any non-yii PSR implementation

**Comment by rob006** [comment]
And what is the advantage of having separate `yiisoft/file-cache` package over having `yiisoft/cache` with file cache (and others) implementation included?

**Comment by alexkart** [comment]
Just more clear structure and separation of responsibilities I think.

**Comment by rob006** [comment]
I think you have confused JavaScript with PHP. :P

I'm not a fan of this refactoring overall. I think that it is overdesigned and focuses on irrelevant things that could be fixed in simpler way. But even I can see some advantages of it. 

I can't say the same about having separate packages for each class - it is just unnecessary complication which results more confusion for users and unreasonable amount of code duplication.

**Comment by samdark** [comment]
I am fine with such separation.

**Comment by rob006** [comment]
> It's not going to cause any duplication.

`normalizeTtl()` is already duplicated in each implementation. It will get even worse if you need to create separate exceptions for each package.

>  As an example, we can take apcu cache.

I'm not talking about apcu cache, but dependency-free implementations. You don't need to specify any extensions in `composer.json` for `FileCache`, `ArrayCache` or `NullCache`. 
Also, apcu is not that obvious. `ApcuCache` does not really work in CLI - you can have this extension disabled for CLI (which will be used by composer) and enabled only in PHP config for web.

**Comment by alexkart** [comment]
@samdark are there repositories for extracting cache implementation?

**Comment by samdark** [comment]
Yes:

- https://github.com/yiisoft/cache-wincache
- https://github.com/yiisoft/cache-file
- https://github.com/yiisoft/cache-memcached
- https://github.com/yiisoft/cache-apcu/

**Pull request merged**